File: scripts/generate-content.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "yaml"

ROOT = File.expand_path("..", __dir__)
SOURCE_DIR = File.join(ROOT, "micro")
OUTPUT_DIR = File.join(ROOT, ".generated-content", "micro")
MONTHS = {
  "Jan" => "01",
  "Feb" => "02",
  "Mar" => "03",
  "Apr" => "04",
  "May" => "05",
  "Jun" => "06",
  "Jul" => "07",
  "Aug" => "08",
  "Sep" => "09",
  "Oct" => "10",
  "Nov" => "11",
  "Dec" => "12"
}.freeze

def slugify(title)
  title
    .split
    .first(6)
    .join(" ")
    .unicode_normalize(:nfkd)
    .encode("ASCII", invalid: :replace, undef: :replace, replace: "")
    .downcase
    .gsub(/['’]/, "")
    .gsub(/[^a-z0-9]+/, "-")
    .gsub(/\A-+|-+\z/, "")
end

def title_from(content, path)
  match = content.match(/^#\s+(.+?)\s*$/)
  raise "Missing H1 heading in #{path}" unless match

  match[1].sub(/\s+\{#.*\}\z/, "").strip
end

def date_from(path)
  parts = path.delete_prefix("#{SOURCE_DIR}/").split("/")
  year, month_name, file_name = parts
  month = MONTHS.fetch(month_name)
  day = file_name.to_i

  format("%s-%s-%02d", year, month, day)
end

FileUtils.rm_rf(OUTPUT_DIR)
FileUtils.mkdir_p(OUTPUT_DIR)

seen_slugs = {}

Dir.glob(File.join(SOURCE_DIR, "**", "*.md")).sort.each do |path|
  content = File.read(path)
  title = title_from(content, path)
  slug = slugify(title)
  raise "Empty slug generated from #{path}" if slug.empty?
  raise "Duplicate slug #{slug.inspect} for #{path} and #{seen_slugs[slug]}" if seen_slugs.key?(slug)

  seen_slugs[slug] = path

  front_matter = {
    "title" => title,
    "slug" => slug,
    "date" => date_from(path),
    "sourcePath" => path.delete_prefix("#{ROOT}/")
  }.to_yaml

  File.write(File.join(OUTPUT_DIR, "#{slug}.md"), "#{front_matter}---\n\n#{content}")
end

puts "Generated #{seen_slugs.length} micro posts"

File: scripts/generate-content.mjs
import { readdirSync, rmSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join, relative } from "node:path";

const root = process.cwd();
const sourceDir = join(root, "micro");
const outputDir = join(root, ".generated-content", "micro");
const months = new Map([
  ["Jan", "01"],
  ["Feb", "02"],
  ["Mar", "03"],
  ["Apr", "04"],
  ["May", "05"],
  ["Jun", "06"],
  ["Jul", "07"],
  ["Aug", "08"],
  ["Sep", "09"],
  ["Oct", "10"],
  ["Nov", "11"],
  ["Dec", "12"],
]);

function slugify(title) {
  return title
    .trim()
    .split(/\s+/)
    .slice(0, 6)
    .join(" ")
    .normalize("NFKD")
    .replace(/[^\x00-\x7F]/g, "")
    .toLowerCase()
    .replace(/['']/g, "")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

function yamlString(value) {
  return `'${value.replaceAll("'", "''")}'`;
}

function titleFrom(content, path) {
  const match = content.match(/^#\s+(.+?)\s*$/m);
  if (!match) throw new Error(`Missing H1 heading in ${path}`);
  return match[1].replace(/\s+\{#.*\}$/, "").trim();
}

function dateFrom(path) {
  const [year, monthName, fileName] = relative(sourceDir, path).split("/");
  const month = months.get(monthName);
  if (!month) throw new Error(`Unknown month ${monthName} in ${path}`);
  const day = Number.parseInt(fileName, 10);
  return `${year}-${month}-${String(day).padStart(2, "0")}`;
}

function markdownFiles(dir) {
  return readdirSync(dir, { withFileTypes: true }).flatMap((entry) => {
    const path = join(dir, entry.name);
    if (entry.isDirectory()) return markdownFiles(path);
    if (entry.isFile() && entry.name.endsWith(".md")) return [path];
    return [];
  });
}

rmSync(outputDir, { force: true, recursive: true });
mkdirSync(outputDir, { recursive: true });

const seenSlugs = new Map();
const paths = markdownFiles(sourceDir).sort();

for (const path of paths) {
  const content = readFileSync(path, "utf8");
  const title = titleFrom(content, path);
  const slug = slugify(title);

  if (!slug) throw new Error(`Empty slug generated from ${path}`);
  if (seenSlugs.has(slug)) {
    throw new Error(`Duplicate slug ${slug} for ${path} and ${seenSlugs.get(slug)}`);
  }

  seenSlugs.set(slug, path);

  const frontMatter = [
    "---",
    `title: ${yamlString(title)}`,
    `slug: ${yamlString(slug)}`,
    `date: ${yamlString(dateFrom(path))}`,
    `sourcePath: ${yamlString(relative(root, path))}`,
    "---",
    "",
  ].join("\n");

  writeFileSync(join(outputDir, `${slug}.md`), `${frontMatter}${content}`);
}

console.log(`Generated ${seenSlugs.size} micro posts`);
